aten/src/ATen/native/SharedReduceOps.h

```diff
@@ -157,15 +150,32 @@ struct MeanOps {
   }
 };
 
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <typename T>
+inline C10_DEVICE T min_impl(T a, T b) {
+  return ::min(a, b);
+}
+
+template <>
+inline C10_DEVICE c10::complex<float> min_impl(c10::complex<float> a, c10::complex<float> b) {
```

Comment:
An implicit assumption in these complex min/max/pow/sqrt functions is that the imaginary part of the complex numbers is 0, since we are always calling `abs` on them first. This assumption is required for numpy compatibility, and it also circumvents the fact that you can't actually perform a true min/max (if inputs' norms are equal but the inputs are not) or a pow (for instance, square root will usually have two correct answers) with complex numbers.

Perhaps it would be better to just take out the real part and operate upon it. If we do that, it might be good to have an assert to check that the imaginary part is 0. It would be a device side assert, which isn't great, but it should never actually get triggered by anything a user can do. The assert would just ensure that developers will only call this on complex numbers that have been normed.

But I think there's an even better solution. Since these AbsMinOps/AbsMaxOps/etc accumulators need to accumulate the absolute values of complex numbers, it would be much better to accumulate them as the downgraded type. This would increase performance and we would avoid needing to use min/max/pow/sqrt functions that accept complex numbers. I'll go ahead and implement this solution, unless anyone has a better suggestion.

cc @mruberry

Reply:
I've updated the norm kernels to use the downgraded scalar type for the accumulators.